jax/interpreters/sharded_jit.py

```diff
@@ -270,3 +336,60 @@ def wrapped(*args, **kwargs):
     return tree_unflatten(out_tree(), out)
 
   return wrapped
+
+
+def _set_sharding_impl(x, partitions):
+  # TODO(skye): can we also prevent this from being called in other
+  # non-sharded_jit contexts? (e.g. pmap, control flow)
+  raise NotImplementedError(
+      "set_sharding() should only be called inside sharded_jit()")
+
+def _set_sharding_translation_rule(c, x_node, partitions):
+  return xb.set_sharding(c, x_node, partitions)
+
+set_sharding_p = core.Primitive("set_sharding")
+set_sharding_p.def_impl(_set_sharding_impl)
+set_sharding_p.def_abstract_eval(lambda x, partitions: x)
+ad.deflinear(set_sharding_p, lambda ct, partitions: (set_sharding(ct, partitions),))
+xla.translations[set_sharding_p] = _set_sharding_translation_rule
+
+def set_sharding(x, partitions: Optional[PartitionSpec]):
```

Comment:
`with_sharding_constraint` ?

Might be a better name since it gives the intuition that information can flow backwards.

Reply:
Good idea, done!